### ISSUE TITLE
Testing by simulated runs as a complement to unit testing to expose otherwise invisible bugs?

### ISSUE THREAD

**Issue by tvleavitt** [triplea-game/triplea#7835]
#7834 highlights that manual testing can fail to expose game-breaking bugs (such as the one it corrects, introduced in #5842 and persistent for 8 months). While current development processes emphasize the need for accompanying unit tests, the nature of the code base is such that even these can't ensure that changes don't alter its behavior as a whole.

Is there a way we could use automated runs of sample games to establish a baseline set of results, and monitor for anomalous outcomes (like the AI building no factories) or other severe deviations in behavior? Maybe run pos2 100x times, compile a set of statistics with ranges of expected output, and then have some kind of automated test that runs it 20x times, not with every build, but on a weekly basis against the current pre-release (and when build a formal "release") and flags any significant deviations? A week should be short enough to narrow down the possibilities for a root cause.

Maybe also filter the "hardAI" logs against a set of expected outputs, and flags any missing elements (no units of type X built)?

Of course, if we make any significant changes to the AI that we regard as improvements, we'll need to establish a new baseline.

Is pos2 a suitable candidate? Do we need to run this test against multiple maps? How many runs is an adequate sample size for a baseline? For a comparison against the baseline? 

What kind of computational resources is this going to take? What would that cost if cloud-hosted? Do we have access to a non-cloud hosted machine for testing purposes? Is there a way we can run a simulation that results in faster execution and lessened resource consumption that just setting all players to HardAI and waiting for a result?

**Comment by trevan** [comment]
The only difference between Fast AI and Hard AI is the battle calculator.  So Fast AI could be a better option than Hard AI.

**Comment by DanVanAtta** [comment]
> Looking at #5842, I note that all the testing done was manual, which points up the severe limitations of that as an approach. Clearly, this needed automated testing.

I agree. I think this stat speaks quite a bit to this topic:

![Screenshot from 2020-10-03 12-56-13](https://user-images.githubusercontent.com/12397753/95000703-ee2b6300-0577-11eb-9d09-1bb0737647e3.png)

- https://codecov.io/gh/triplea-game/triplea/tree/master/game-core/src/main/java/games/strategy/triplea


What you're proposing here is a pretty large test structure, a hybrid of an acceptance, regression and system test. I'm very nervous that we would fall into the integrated testing trap. This is a good video on the topic:
- https://www.youtube.com/watch?v=VDfX44fZoMc "J B Rainsberger Integrated Tests Are A Scam HD"

On the other side of that, mocks are a testing code-smell as well but often necessary (perhaps saying more specifically overly heavy use of mocks). So there is a balance between to be struck therefore on how large of a "unit" should be when it comes to testing. Having the "unit" size be a method is too small as it creates pervasive use of mocking which then has the testing to just test the implementation and not the behavior. If the pieces are chosen to be a good size, with some design, then the strategy of JB Rainsberger works very well as you can then do a proof by induction that the full system would work.

Perhaps that is an aside, as our current situation is almost zero testing of the AI code. I suspect a good chunk of the 1% of test coverage is incidental coverage from other tests.

---------

To speak to the strategy proposed, it would be a good thing for checking/finding errors. I don't think we'd want to compute statistics and base our test result on that. Those tests would not be deterministics and would not pinpoint the location of a problem. A failure would not necessary indicate a regression has been introduced in the system and with no real indicator of where the problem would lead to failed builds and hours of debugging for problems that were potentially already in the system. 

The other aspect, mentioned by JB Rainsberger, the surface area of what we need to test is the combination of all code paths of the system under test. This has combinatoric growth. There will always therefore be gaps in what we can detect through such a strategy. Hence my suggestion that this type of test would be good to flag hard errors (things like NPEs).

To really test the AI code we need to modularize it more and create decision modules where given an input game state we get an output decision. Currently the AI code is very involved and does not have this structure at all. I think next we'd have to start a system where we can create game states without need to parse a full XML to get to that game state. This requires more core changes to the game engine and perhaps completion of the save-game serialization project. The way that game is structure today, because of serialization, almost all domain objects require a reference to a `GameData` object, which is typically only available if you parse a full XML. Hence, if you want to create domain objects for an input, you are stuck creating `GameData` objects, and hence for such a limited scenario you need to parse a full map. This greatly limits our ability to feed in input scenarios.

One thought, if we can make the AI deterministic and inject the values for random seeds, we could build a decent regression test suite by validating that the AI plays a given game the same way. I think this would be quite expensive to build though.

So in summary, a holistic testing strategy is a lot of work:
- the answer of going from effectively 0% test coverage is not necessary to add in very expensive system tests (though those system tests can still have a place, but not for checking for specific AI behavior).
 - to test the AI effectively we need to:
  - decouple domain objects from a `GameData` so we can set up scenarios using domain objects without needing to create those objects via XML. It is far too many scenarios, we are talking hundreds to then maintain hundreds of XMLs
  - modularize the AI code so that we can feed in sets of domain objects and validate we get a correct computation output


In the short term, I think this is all emphasis on proper refactoring techniques and working with legacy code. We have had a lot of trouble to enforce this as writing tests for one line fixes is not a popular thing to ask for. What a person *can* do when modifying code is:
- extract the code that will be modified to an object or a static method that has an input/output interface that can be tested
- write a test for this
- check *that* in
- update the extracted code that is now tested.

The problem with this approach is that restructuring later can be quite difficult. If two modules are merged and/or split, then the testing can be invalidated. So there is care that needs to be taken to create somewhat the right modules.  Typically having a high level test is another strategy to complement the low level tests to mitigate this problem, but then we run into the problems needed for more holistic testing of the AI code mentioned above.

TLDR:
- there is almost zero coverage of the AI code, it and the core game engine code are tied into a knot, we won't get the test efficacy we want through large integration testing alone, and to get a proper testing strategy in place involves quite a lot of work and depends on projects that are somewhat already in motion.

**Comment by DanVanAtta** [comment]
> What kind of computational resources is this going to take? What would that cost if cloud-hosted? Do we have access to a non-cloud hosted machine for testing purposes? Is there a way we can run a simulation that results in faster execution and lessened resource consumption that just setting all players to HardAI and waiting for a result?

We could simply use travis. We'd want to balance build time. As mentioned above, these test would be great for flagging major errors, but not for specific behavior.

**Comment by DanVanAtta** [comment]
There is quite a bit of history and technical background here, sorry for such a long-winded set of answers @tvleavitt . You are right to raise such a question and the lack of testing is arguably one of the biggest downfalls of the TripleA codebase. I can't find the reference for it anymore, but someone wrote a masters paper on the impact of automated testing on software projects and included TripleA. Their conclusion was that TripleA scored very highly for poor testing which correlated to slow development. They essentially predicted development would come to a stand-still, and to some extent they were right. Between 2014 and arguably now, there have not really been that many major changes.

Fundamentally I think the answer is a cultural one. That we need to require changes to pay back the massive debt and build in testing. The codebase is in knots so this is not always practical, time effective, or even always helpful. In the medium term, the strategy is roughly two-fold:
- fix the coupling of all domain objects back to XML (save-game project)
- partition the code with sub-projects and then further sub-divide into modules

That would probably allow us to write effective tests.  How to fix the culture problem is an open question and difficult to justify until we can be really effective with tests and there has been a lot of back and forths around asking that we build tests into code we change. So, we find ourselves in quite the catch-22:
- to untangle the knots in the code we need tests
- to write effective tests, we need fewer knots in the code
- to ask the team to write tests that would help untangle the knots, we need to be able to write effective tests

**Comment by DanVanAtta** [comment]
TL;DR,

> Testing by simulated runs as a complement to unit testing to expose 

When it comes to the AI code, there basically are not unit tests. I'm afraid the premise of this issue is faulty. Though the simulated runs could have other uses. to get there we need to be able to add AI to headless games. I once gave that a try and it worked but the AI did not make any moves. The "smoke-testing" project we have is perfect for this and would be a natural progression. That project fires up a lobby, a bot, connects the two, and joins the bot. If we can just go a bit further and have the bot set all AI players and launch a game, we would be well on our way.

**Comment by tvleavitt** [comment]
@DanVanAtta My takeaway from your comments is that you see a use case for this, conceptually, albeit limited, thus I was not entirely off in suggesting it. You raised the idea of integrating this into Travis (as a means of detecting NPE's and other major errors), and I think you're outlining a path to implementation (outside of Travis?) in the above post.

What actionable items can we derive from this conversation? Do we create an issue? Define a kanban "project" (which I'm using in another context on GitHub) and then extract issues out of that?

On the original topic, more generally, I understand the overall limitations of this approach. I was thinking of it less as a way of directly generating issues and more as a backup method of flagging major behavioral changes. If I'm running a chip fab, and I notice my error rate go from 1% to 5% in two weeks, that's not going to tell me the precipitate cause, but it's a signal that *something* has changed, somewhere, and needs to be looked into. If it raises an issue inside of two weeks from the change, perhaps it would tickle a thought in the mind of one of the developers as to what they might have done, perhaps a recollection of a fleeting thought that maybe this might have a side effect, that was not pursued.

But, again, everything is a balance, and I'll defer to the judgment of those actively writing the code, and who have the technical chops to implement a solution that I lack.

**Comment by DanVanAtta** [comment]
As part of release testing we do run all AI games to see if they generate errors. I think we are essentially now talking about automating that and how to see that happen. 

We are currently not good at tracking the roadmap. Maybe we could make extensive use of projects to track development initiatives? If so, then a project would be the right thing to do. Otherwise we could add a line-item in a document, we do have a future-initiatives document, that would detail this. My preference is probably for the former. I do think an issue with a special label or something would get lost. I think I like the idea to create projects for such relatively important initiatives.
